File: Tests/ConnectionAttributionCoverageTests.swift
import XCTest

@testable import MihomoMeter

final class ConnectionAttributionCoverageTests: XCTestCase {
  func testCountsEachConnectionOnceAndUpgradesAvailability() {
    var tracker = ConnectionAttributionCoverageTracker()

    XCTAssertEqual(
      tracker.consume([
        connection("first", hostname: true, application: false),
        connection("second", hostname: false, application: true),
      ]),
      ConnectionAttributionCoverage(
        proxyConnectionCount: 2,
        hostnameIdentifiedCount: 1,
        applicationIdentifiedCount: 1,
        fullyIdentifiedCount: 0
      )
    )

    XCTAssertEqual(
      tracker.consume([
        connection("first", hostname: false, application: true)
      ]),
      ConnectionAttributionCoverage(
        proxyConnectionCount: 2,
        hostnameIdentifiedCount: 1,
        applicationIdentifiedCount: 2,
        fullyIdentifiedCount: 1
      )
    )
  }

  func testResetRemovesInMemoryConnectionIdentifiers() {
    var tracker = ConnectionAttributionCoverageTracker()
    _ = tracker.consume([connection("connection", hostname: true, application: true)])

    tracker.reset()

    XCTAssertEqual(tracker.consume([]), .empty)
  }

  private func connection(
    _ id: String,
    hostname: Bool,
    application: Bool
  ) -> ConnectionTrafficSample {
    ConnectionTrafficSample(
      id: id,
      bytes: .zero,
      chains: ["Proxy"],
      metadata: ConnectionMetadata(
        hostname: hostname ? "example.com" : nil,
        applicationName: application ? "Example" : nil
      )
    )
  }
}

File: platform/windows/MihomoMeter.Windows.Tests/ConnectionAttributionCoverageTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MihomoMeter.Windows.Core.Domain;

namespace MihomoMeter.Windows.Tests;

[TestClass]
public sealed class ConnectionAttributionCoverageTests
{
    [TestMethod]
    public void CountsEachProxyConnectionOnceAndUpgradesAvailability()
    {
        var tracker = new ConnectionAttributionCoverageTracker();

        var initial = tracker.Consume(
        [
            Connection("first", hasHostname: true, hasApplication: false),
            Connection("second", hasHostname: false, hasApplication: true),
        ]);
        var upgraded = tracker.Consume(
        [
            Connection("first", hasHostname: false, hasApplication: true),
        ]);

        Assert.AreEqual(
            new ConnectionAttributionCoverage(2, 1, 1, 0),
            initial);
        Assert.AreEqual(
            new ConnectionAttributionCoverage(2, 1, 2, 1),
            upgraded);
        Assert.AreEqual(0.5, upgraded.HostnameRate);
        Assert.AreEqual(1.0, upgraded.ApplicationRate);
        Assert.AreEqual(0.5, upgraded.FullyIdentifiedRate);
    }

    [TestMethod]
    public void ResetRemovesInMemoryConnectionIdentifiers()
    {
        var tracker = new ConnectionAttributionCoverageTracker();
        _ = tracker.Consume([Connection("connection", true, true)]);

        tracker.Reset();

        Assert.AreEqual(ConnectionAttributionCoverage.Empty, tracker.Consume([]));
    }

    private static ConnectionTrafficSample Connection(
        string id,
        bool hasHostname,
        bool hasApplication)
    {
        return new ConnectionTrafficSample(
            id,
            TrafficBytes.Zero,
            ["Synthetic Proxy"],
            new ConnectionMetadataAvailability(hasHostname, hasApplication));
    }
}
